### PR TITLE
[[nodiscard]] messages

### LINKED ISSUE #271: STL: Consider using P1301R4 [[nodiscard("message")]]
WG21-P1301 `[[nodiscard("message")]]` has been implemented for VS 2019 16.5 and has [also been implemented](https://clang.llvm.org/cxx_status.html) by Clang 9.

We should consider using this in the STL. We have a few criteria for marking functions as `[[nodiscard]]`: pure observer (most common), resource leak, incorrect behavior (e.g. `remove_if()`). We could have macros for each of these reasons.

### LINKED ISSUE #1742: <mutex>: lock_guard can emit nodiscard warnings for valid code
Reported as DevCom-1368582, caused by #1495 (unrelated to #1514):

```cpp
#include<mutex>

struct Mutexed {
  mutable std::mutex m;
  int i;
};

struct MyClass {
  int i;
  explicit MyClass(Mutexed const& m) :
    i{ (std::lock_guard{ m.m }, m.i) } // warning C4834: discarding return value of function with 'nodiscard' attribute
  {}
};
```

This code is valid and reasonable - it's not doing anything terribly weird. There are other ways to write the code (including ways that don't involve assigning in the constructor body - e.g. a helper function could lock the `mutex` and read the `int`), but there's nothing inherently wrong with this way.

Our usual policy for `[[nodiscard]]` is to strongly prefer avoiding false positives - to the point where if 10% of uses are valid and 90% are bogus, we would prefer to not warn. (`unique_ptr::release()` is a good example where we don't warn.)

We need to decide what to do in this case. My feeling is that while this use of the comma operator is valid, it doesn't seem common at the 10% level (my general impression is that usage of the comma operator is quite rare, as most programmers are unfamiliar with the rule that temporaries live until the end of the full-expression). I also feel that it is *extremely hazardous* to mistakenly discard a `lock_guard`, because the code will appear to work most of the time. (Whereas a discarded `unique_ptr::release()` is a memory leak, which can be noticed later.) As we don't have a way to tell the compiler to treat LHS comma subexpressions differently from expression-statements, I think we should resolve this as by design.

Fortunately, there is a workaround that allows this code to continue to compile without the need for push-disable-pop. Casting the LHS to `(void)` is sufficient to avoid the warning: https://godbolt.org/z/b5f6vz

### PR DESCRIPTION
**Pull request by AlexGuteniev**
Fixes #271
Fixes #1742

### CODE REVIEW

**Comment by fsb4000** [comment]
You can also add: "Fixes #1742"

**Comment by AlexGuteniev** [comment]
> Oops, I was wrong

Why, it still fixes #1742 where possible

**Comment by fsb4000** [comment]
>Why, it still fixes #1742 where possible

You are correct.
I tested on godbolt: https://godbolt.org/z/r1fPM3b9K
So `__has_cpp_attribute(nodiscard) >= 201907L` is true for `/std:c++14`

File: stl/inc/yvals_core.h [hunk 1]
```diff
@@ -414,14 +414,100 @@
...
-#define _NODISCARD_CTOR _NODISCARD
+#define _NODISCARD_CTOR           _NODISCARD
+#define _NODISCARD_CTOR_MSG(_Msg) _NODISCARD_MSG(_Msg)
 #else
 #define _NODISCARD_CTOR
+#define _NODISCARD_CTOR_MSG(_Msg)
 #endif
 
+#define _NODISCARD_PURE_ALGORITHM                                                                                   \
+    _NODISCARD_MSG("This algorithm  is not intended to have side effects; it is not useful to call this algorithm " \
```
Comment (sam20908) [inline stl/inc/yvals_core.h:436]:
```suggestion
    _NODISCARD_MSG("This algorithm is not intended to have side effects; it is not useful to call this algorithm " \

```

**Review by sam20908 — CHANGES_REQUESTED** [review]

File: stl/inc/yvals_core.h [hunk 2]
```diff
@@ -414,14 +414,119 @@
...
+        "current statement, using comma operator, please use cast to void to suppress this warning)"                 \
+        "Alternatively, define _NODISCARD_LOCK_SUPPRESS.)")
+
+#endif // ^^^ !defined(_NODISCARD_LOCK_SUPPRESS) ^^^
+
+#define _NODISCARD_CTOR_THREAD \
+    _NODISCARD_CTOR_MSG("Thread is not joined or detached, terminate will be called at the end of the statement")
+
+#define _NODISCARD_CTOR_JTHREAD \
+    _NODISCARD_CTOR_MSG("Thread is implicitly joined at the end of the statement, is this an intention?")
```
Comment (AlexGuteniev) [inline stl/inc/yvals_core.h:524]: While I agree `std::thread` constructor should be `[[nodiscard]]`, I'm questioning it for `std::jtrhead` constructor.

Comment (AlexGuteniev) [reply stl/inc/yvals_core.h]: It can be suppressed with `jthread(...).join()`

**Comment by StephanTLavavej** [comment]
There's a merge conflict in `<vector>`.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks for this major diagnostic improvement :heart_eyes_cat: - and apologies for taking so very long to review this! I've pushed a bunch of changes, renaming/rephrasing stuff and updating a few more locations to use the message macros. Please let me know if anything looks undesirable. (I was probably most aggressive with rephrasing the `assume_aligned` message; I thought it was better to follow your `launder` pattern and not describe in too much detail how the compiler may or may not optimize.)

**Comment by AlexGuteniev** [comment]
> Please let me know if anything looks undesirable.

I intentionally did not place full stop at the end of each message for consistency with other ranges. Consider this program:
```C++
[[nodiscard("don't discard. the value is useful")]]
bool lock()
{
    return false;
}

int main()
{
    int i;
    lock();
}
```

The warnings (`/W4`):
```
ConsoleApplication14.cpp(19,9): warning C4858: discarding return value: don't discard. the value is useful
ConsoleApplication14.cpp(18,9): warning C4101: 'i': unreferenced local variable
```

Do you still want to have it?

**Comment by AlexGuteniev** [comment]
I'm fine with new `assume_aligned`. I also think `launder` message could also mention "a _potential_ effect".
At least I can't  confirm any evidence it might have some real effect on MSVC.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
(capturing Discord discussion here for the record)

The compiler definitely prefers short phrases without periods for most of its warning messages. However, longer messages (typically static analysis warnings) are sometimes full sentences with periods, like https://godbolt.org/z/v9ansc3fa :

```
<source>(3) : warning C28251: Inconsistent annotation for 'new': this instance has no annotations.
See c:\users\containeradministrator\appdata\local\temp\compiler-explorer-compiler2022728-2624-10iodxs.ufwa\predefined c++ types (compiler internal)(33).
The first user-provided annotation on this built-in function is at line c:\data\msvc\14.33.31629\include\vcruntime_new.h(48). 
```

And our `[[deprecated("message")]]` explanations, which are the closest precedent for `[[nodiscard("message")]]`, have periods: https://github.com/microsoft/STL/blob/8ca7bd3c002d383940de9abe6379721e96eb12e3/stl/inc/yvals_core.h#L935-L939

Thus while there are arguments for both ways, we've concluded that keeping the periods would be best.

**Comment by StephanTLavavej** [comment]
I'm *speculatively* mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

File: stl/inc/yvals_core.h [hunk 3]
```diff
@@ -488,6 +502,110 @@
...
+                        "If this is intentional, you can add '.join()' to suppress this warning. "         \
+                        "Otherwise, this 'std::jthread' should be stored in a variable.")
+
+#define _NODISCARD_ASYNC                                                                                           \
+    _NODISCARD_MSG("The result of 'std::async' should be stored in a variable. If the return value is discarded, " \
+                   "the temporary 'std::future' is destroyed, waiting for an async result or evaluating "          \
+                   "a deferred result, thus defeating the purpose of 'std::async'.")
+
+#define _NODISCARD_GET_FUTURE                                                                              \
+    _NODISCARD_MSG("Getting the future more than once or not satisfying the obtained future will throw a " \
```
Comment (CaseyCarter) [inline stl/inc/yvals_core.h:606]: I have no idea what "satisfying the obtained future" means.

Comment (CaseyCarter) [reply stl/inc/yvals_core.h:606]: Is this trying to say "satisfying the corresponding `promise`"? If so, it's confusing to apply this to `packaged_task::get_future`.

Comment (CaseyCarter) [reply stl/inc/yvals_core.h:606]: I'm going to approve to avoid resetting testing, and address this comment in a followup.

**Review by CaseyCarter — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks again for enhancing the STL's favorite warning! :heart_eyes_cat: :warning: :rocket:

**Pull request merged**